core/controllers/load_functions.py

```diff
@@ -0,0 +1,219 @@
+import pandas as pd
+from sqlalchemy import desc, func
+
+from core.const import SUBMISSION_ID_FORMAT
+from core.controllers.mappings import DataMapping
+from core.db import db
+from core.db.entities import Organisation, Project, Submission
+from core.db.queries import (
+    get_programme_by_id_and_previous_round,
+    get_programme_by_id_and_round,
+)
+
+
+def load_programme_ref(workbook: dict[str, pd.DataFrame], mapping: DataMapping):
+    """Loads data into the 'Programme_Ref' table.
+
+    If the programme exists in a previous round, the incoming programme's information supersedes the existing programme.
+    Data is added to the session via this method, but not committed.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    """
+    reporting_round = int(workbook["Submission_Ref"]["Reporting Round"].iloc[0])
+    programme_id = workbook["Programme_Ref"]["Programme ID"].iloc[0]
+    # if added before or during this round, get the programme entity to merge (update)
+    programme_exists_previous_round = get_programme_by_id_and_previous_round(programme_id, reporting_round)
+
+    worksheet = workbook[mapping.table]
+    models = mapping.map_data_to_models(worksheet)
+
+    if programme_exists_previous_round:
+        # Set incoming model pk to match existing DB row pk (this record will then be updated).
+        programme_to_merge = models[0]
+        programme_to_merge.id = programme_exists_previous_round.id
+        db.session.merge(programme_to_merge)
+    else:
+        db.session.add_all(models)
```

Comment:
This results in ```sqlalchemy.orm.exc.UnmappedInstanceError: Class 'builtins.list' is not mapped```

Reply:
I suppose we'd have to do ```db.session.add(models[0])```? Not sure if that's any clearer or not

Reply:
I think unpack the Programme instance directly after grabbing the attribute from mapping. Since it also gets unpacked with `programme_to_merge = models[0]` inside the if block. So just move that line outside the block and rename it to something more generic? ie:

```
    models = mapping.map_data_to_models(worksheet)
    programme = models[0]

    if programme_exists_previous_round:
        programme.id = programme_exists_previous_round.id
        db.session.merge(programme)
    else:
        db.session.add(programme )
```